Solving Corrupt Cards #15222 caused by bad entries in json columns

adam-james-v opened the pull request:
fixes: #15222

This PR is the starting point to address 'corrupt cards'. This is kind of one bug, as the root cause is invalid json in the app-db, causing the frontend to fail (in the case of no query-dataset being available), and the backend to fail (in the case of un-normalizeable visualization-settings).

To reproduce, in a Clojure dev environment:

1. create and save a card and remember its id.
2. in the REPL, `(toucan.db/update! Card card-id :dataset_query "")`
 - note that if you do `(:dataset_query (toucan.db/select-one Card :id card-id))` you should see `nil`. This is because `cheshire.core/parse-string` will return `nil` for blank strings.
 - also note that you cannot update the :dataset_query key with `nil` directly, as there is a not-null constraint. So, the blank string basically circumvents this constraint, and is the root of our issue, I believe.
3. If you reload the question in your browser, you will see 'something's gone wrong'. You can further try to reproduce based on the discussion in #15222, but this is a quick way to see this failure.

1. create and save a card and remember its id.
2. in the REPL, `(toucan.db/update! Card card-id :visualization_settings (cheshire.core/parse-string "{\"column_settings\":\"{\\\"(keyword [\\\\\\\"name\\\\\\\" \\\\\\\"ID\\\\\\\"])\\\":{\\\"number_separators\\\":\\\".\\\"}}\"}\n" keyword))`
3. Reload the question in the browser. You should see 'something went wrong' again.

I have created a post-select in the card api that will replace the :dataset_query with a 'dummy query' in the situation where :dataset_query has a `nil` value. This will not recover the card's query, and I'm pretty sure the card is effectively lost, but it should not break the frontend anymore, as the question can load properly, from the app's perspective. _I do not know if there's a way to recover the query_. I am very open to suggestions or improvements on this approach.

I have wrapped the visualization_settings normalization in a try catch that will log the failure but return the empty map `{}`. This allows the app to load, but the visualization settings will be lost. Upon opening and saving the question, new, valid viz settings will be written to the app-db. The original viz-settings will be lost, but the card is still safe otherwise, and viz-settings are (somewhat) easier to recreate.

adam-james-v committed:
Solving Corrupt Cards #15222 caused by bad entries in json columns

This PR is the starting point to address 'corrupt cards'. This is kind of one bug, as the root cause is invalid json
in the app-db, causing the frontend to fail (in the case of no query-dataset being available), and the backend to
fail (in the case of un-normalizeable visualization-settings).

To reproduce, in a Clojure dev environment:

1. create and save a card and remember its id.
2. in the REPL, `(toucan.db/update! Card card-id :dataset_query "")`
 - note that if you do `(:dataset_query (toucan.db/select-one Card :id card-id))` you should see `nil`. This is
 because `cheshire.core/parse-string` will return `nil` for blank strings.
 - also note that you cannot update the :dataset_query key with `nil` directly, as there is a not-null constraint. So,
 the blank string basically circumvents this constraint, and is the root of our issue, I believe.
3. If you reload the question in your browser, you will see 'something's gone wrong'. You can further try to reproduce
based on the discussion in #15222, but this is a quick way to see this failure.

1. create and save a card and remember its id.
2. in the REPL, `(toucan.db/update! Card card-id :visualization_settings (cheshire.core/parse-string "{\"column_settings\":\"{\\\"(keyword [\\\\\\\"name\\\\\\\" \\\\\\\"ID\\\\\\\"])\\\":{\\\"number_separators\\\":\\\".\\\"}}\"}\n" keyword))`
3. Reload the question in the browser. You should see 'something went wrong' again.

I have created a post-select in the card api that will replace the :dataset_query with a 'dummy query' in the
situation where :dataset_query has a `nil` value. This will not recover the card's query, and I'm pretty sure the card
is effectively lost, but it should not break the frontend anymore, as the question can load properly, from the app's
perspective. _I do not know if there's a way to recover the query_. I am very open to suggestions or improvements on
this approach.

I have wrapped the visualization_settings normalization in a try catch that will log the failure but return the empty
map `{}`. This allows the app to load, but the visualization settings will be lost. Upon opening and saving the
question, new, valid viz settings will be written to the app-db. The original viz-settings will be lost, but the card
is still safe otherwise, and viz-settings are (somewhat) easier to recreate.

adam-james-v commented on src/metabase/models/card.clj:
```diff
@@ -305,6 +305,13 @@
+  (cond-> card
+    (nil? dataset-query) (assoc :dataset_query {:database (:database_id card)
+                                                :type     :native
+                                                :native   {:query "select 'this query broke, sorry' x"}})))
```
I don't love this approach really. I'd rather have things actually fail and not break the app, but this solution at least gets a corrupt card to 'load' (for some sense of the word).

If there is a way to reconstruct the actual query, that would be much cooler.

Obviously, I think there needs to be a way to more robustly prevent invalid json from entering the db in the first place, so that we don't need a weird post-select solution like this.

adam-james-v commented on src/metabase/models/interface.clj:
```diff
@@ -157,9 +160,21 @@
+    (catch Throwable e
+      (log/error e (tru "Unable to normalize:") "\n"
+                 (u/pprint-to-str 'red viz-settings))
+      {})))
```
If you pass 'nil' instead of an empty map, the app will load, but the app viz buttons don't cause changes. For example, I tried to change a bar chart's colors, but it would not change until I saved the question (thus saving valid viz-settings to the app-db), after which it worked fine.

qnkhuat committed:
add tests and maybe fix tests?

tsmacdonald commented on test/metabase/models/card_test.clj:
```diff
@@ -377,3 +377,23 @@
       (is (= #{["Card" (:id card1)]}
              (serdes.base/serdes-descendants "Card" (:id card2)))))))
+
+(deftest corrupted-cards-will-fix-itself
```
```suggestion
(deftest corrupted-cards-will-fix-themselves
```

:wink:

tsmacdonald commented on test/metabase/models/card_test.clj:
```diff
@@ -377,3 +377,23 @@
+        (is (= {:database (mt/id),
+                :native   {:query ""},
+                :type     :native}
+               (:dataset_query (db/select-one Card :id id1))))
```
nit, you could `db/select-one-field` to save calling `:dataset_query` separately

tsmacdonald reviewed (commented):
I'm open to changing to a :heavy_check_mark: after discussion, but having just futzed with JSON migrations it makes sense to me to leave `post-select` alone and manage things with the type converter. The main benefit is that it keeps the logic in one place that's very close to the DB, but it also has the I-think-it's-a-benefit that appropriate dummy values will be persisted in the DB and not just converted at select time.

Heck, it occurs to me that if I got my migration thing merged you could write this as a JSON migration, which would be kinda nifty. You could set the expected version to `2` but not persist the version upgrades (i.e. never call `json-migration/update-version`) and that would ensure that it ran the conversion code every time. <sub>(If nothing else it feels like an endorsement of my API design that it solves this similar-but-different problem.)</sub>

-----

All that said, given the world we live in I agree that we're stuck fixing mysteriously-broken data and this is similar to code I'd write myself. But it feels bad that we don't know what's inserting `""` queries into the DB in the first place, and I'd feel better about this if we fixed the root cause in addition to putting a band-aid over the symptoms.

It'd be great if we could just get rid of all the invalid cards, but I suspect customers would not be happy about that.

adam-james-v committed:
Update test/metabase/models/card_test.clj

Co-authored-by: Tim Macdonald <tim@metabase.com>

adam-james-v committed:
Dataset validation w/ schema. Use in Card API and Card model pre-insert/update

Move the invalid check for datasets out of the post-select and into the card API, since it is the frontend that breaks
when the dataset is invalid or nil. We still don't want it to happen, so there's a pre-insert/update to catch this.

adam-james-v committed:
Open the schema to allow for other keys (eg. database key)

adam-james-v committed:
Tests use {} in dataset_query, so I suppose its valid?

adam-james-v committed:
Actually, post-select might be the best place to catch things.

adam-james-v commented:
@tsmacdonald It’s still not going to pass the tests, but things are stepping forward. One issue is that if I now validate the query on a pre-insert/update, we have tests failing because there’s not a valid default key configured. If you’re looking for something to do, I’d love help with that.

As a worst-case scenario, to get this merged tomorrow ASAP, we could also remove the pre-insert/update check and just make sure the post-select is working, and do a follow-up PR that puts the validation in place after the release.

tsmacdonald committed:
Clean up some tests that used an invalid dataset_query

tsmacdonald commented:
@adam-james-v my two big changes were changing the default value of `dataset_query` use in `with-temp` test helpers so that things would actually be valid, and also futzing with the card test you wrote so that it would actually test the `post-select` behavior. It'd be a stronger test if we inserted invalid values into the DB, but I didn't see a simple way to do that. I thought about writing a `with-temp*` variant called `with-temp-simple*` or something that called `db/insert-simple!` instead of `db/insert!`, but it felt like maybe that should live in Toucan instead of our own test util and I decided it wasn't worth dealing with right now.

tsmacdonald committed:
Fix alert tests :wrench:

tsmacdonald committed:
WIP on more test fixes

adam-james-v committed:
Start fixing tests, now that :dataset_query is validated w. mbql.s/Query

adam-james-v committed:
Some quick fixes for dashboard tests. not perfect, but they work

adam-james-v committed:
More test fixes

adam-james-v committed:
Continue fixing tests

adam-james-v committed:
More test fixes

adam-james-v committed:
Give Temp Cards a valid dataset_query by adding :type key
